README.md
DOG LCD Python for Raspberry Pi
===============================

Special thanks to Eberhard Fahle who created the Arduino DogLcd library
which I have ported to create this module.

Intro
-----

This library supports the DOGM081E, DOGM163E and DOGM183E displays with ST7036 driver IC over
the 4-wire SPI serial interface ( SI: Serial Input, CLK: Clock, RS: Register-Select and CSB: Chip-Select ).

You must connect at least SI, CLK and RS but can add multiple displays using an extra pin for each CSB.

Refer to the DOG LCD datasheet for details on wiring and support components for 3.3v SPI.

http://www.lcd-module.com/eng/pdf/doma/dog-me.pdf

Note: This library has only been tested against the DOGM163E-A

Getting Started
---------------

Install the doglcd module and use as follows:

    import doglcd
	
	CONTRAST = 0x28

    lcd = doglcd.DogLCD(SI, CLK, RS, CSB, RESET, BACKLIGHT)
    lcd.begin(doglcd.DOG_LCD_M163, CONTRAST)

	lcd.clear()
	lcd.home()

	lcd.write('Hello World!')


Usage
=====
These are the doglcd functions you can use,
anything not listed here may have unexpected results

reset()
-------
Reset the LCD

clear()
-------
Clear the LCD

home()
------
Relocate cursor to 0,0

noCursor() / cursor()
---------------------
Turn the cursor off / on

noDisplay() / display()
-----------------------
Turn the display off / on

noBlink() / blink()
-------------------
Turn blinking cursor off / on

scrollDisplayLeft() / scrollDisplayRight()
------------------------------------------
Manually scroll the display 1 character to the left or right

autoScroll() / noAutoscroll()
-----------------------------
Toggles on / off automatic scrolling

createChar(index,arr_char)
--------------------------
Create a new char at index ( 0 to 7 )
using an array of 8 5-bit numbers ( 0 to 31 )

Empty char:
    [0,0,0,0,0,0,0,0]

Full char:
	[31,31,31,31,31,31,31,31]

rightToLeft() / leftToRight()
-----------------------------
Change text direction

writeChar(int)
--------------
Write a valid single char from 0 to 255

write(string/int)
-------------
Write a string or number to the LCD






Hardware setup
==============

DOG LCD Data Pinout
-------------------

TOP LEFT
* 40 : RST, Reset : Pull to VCC
* 39 : RS, Register Select ( Low = Command / High = Data ) -> Pi
* 38 : CSB, Chip Select -> Pi
* 37 : RW ( Low = Write / High = Read ) : Pull to GND
* 36 : E, Enable : Pull to VCC
* 35 : D0 : Pull to VCC
* 34 : D1 : Pull to VCC
* 33 : D2 : Pull to VCC
* 32 : D3 : Pull to VCC
* 31 : D4 : Pull to VCC
* 30 : D5 : Pull to VCC
* 29 : D6 : CLK, Clock -> Pi
* 28 : D7 : SI, Serial Input -> Pi
* 27 : GND
* 26 : VCC
* 25 : Voltage Booster IN : Pull to VCC
* 24 : Voltage Booster OUT : Bridge to 25 with 0.1uF Cap
* 23 : PSB ( Low = Serial / High = Parallel ) : Pull to GND
* 22 : Voltage Booster +
* 21 : Voltage Booster - : Bridge to 22 with 0.1uF Cap

CHANGELOG.txt
Change Log
==========

1.0-dev
-------
- Initial development/beta release
- Only tested against DOGM163E-A @ 3.3v

examples/test.py
#!/usr/bin/env python

import doglcd, datetime, time

# __init__(self, lcdSI, lcdCLK, lcdRS, lcdCSB, pin_reset, pin_backlight):
# pin   wpi   bcm
# SI      2   27 rev 2, 21 rev 1
# CLK     3   22
# RS      5   24
# CSB     4   23
# RST    -1
# BL     -1
#
lcd = doglcd.DogLCD(21,22,24,23,-1,-1)
lcd.begin(doglcd.DOG_LCD_M163, 0x28)
lcd.clear()
lcd.home()
lcd.write(chr(0) + 'OMG! Such time' + chr(0))
lcd.setCursor(0,2)
lcd.write(chr(1) + chr(4) + ' Very Wow! ' + chr(3) + chr(2) + chr(5))


pirate = [
	[0x00,0x1f,0x0b,0x03,0x00,0x04,0x11,0x1f],
	[0x00,0x1f,0x16,0x06,0x00,0x08,0x03,0x1e],
	[0x00,0x1f,0x0b,0x03,0x00,0x04,0x11,0x1f],
	[0x00,0x1f,0x05,0x01,0x00,0x02,0x08,0x07]
]

heart = [
	[0x00,0x0a,0x1f,0x1f,0x1f,0x0e,0x04,0x00],
	[0x00,0x00,0x0a,0x0e,0x0e,0x04,0x00,0x00],
	[0x00,0x00,0x00,0x0e,0x04,0x00,0x00,0x00],
	[0x00,0x00,0x0a,0x0e,0x0e,0x04,0x00,0x00]
	]

raa = [
	[0x1f,0x1d,0x19,0x13,0x17,0x1d,0x19,0x1f],
	[0x1f,0x17,0x1d,0x19,0x13,0x17,0x1d,0x1f],
	[0x1f,0x13,0x17,0x1d,0x19,0x13,0x17,0x1f],
	[0x1f,0x19,0x13,0x17,0x1d,0x19,0x13,0x1f]
	]

arr = [
	[31,14,4,0,0,0,0,0],
	[0,31,14,4,0,0,0,0],
	[0,0,31,14,4,0,0,0],
	[0,0,0,31,14,4,0,0],
	[0,0,0,0,31,14,4,0],
	[0,0,0,0,0,31,14,4],
	[4,0,0,0,0,0,31,14],
	[14,4,0,0,0,0,0,31]
]

char = [
	[12,11,9,9,25,25,3,3],
	[0,15,9,9,9,25,27,3],
	[3,13,9,9,9,27,27,0],
	[0,15,9,9,9,25,27,3]
]

pacman = [
	[0x0e,0x1f,0x1d,0x1f,0x18,0x1f,0x1f,0x0e],
	[0x0e,0x1d,0x1e,0x1c,0x18,0x1c,0x1e,0x0f]
]

def getAnimFrame(char,fps):
	return char[ int(round(time.time()*fps) % len(char)) ]

while 1:
	lcd.createChar(0,getAnimFrame(char,4))
	lcd.createChar(1,getAnimFrame(arr,16))
	lcd.createChar(2,getAnimFrame(raa,8))
	lcd.createChar(3,getAnimFrame(pirate,2))
	lcd.createChar(4,getAnimFrame(heart,4))
	lcd.createChar(5,getAnimFrame(pacman,3))
	lcd.setCursor(0,1)
	t = datetime.datetime.now().strftime("%H:%M:%S.%f")
	lcd.write(t)
	

examples/lr.py
import doglcd, time, datetime, math

BORDER_TOP = 0
BORDER_BOTTOM = 1
BORDER_LEFT = 2
BORDER_RIGHT = 3
BORDER_TL = 4
BORDER_TR = 5
BORDER_BL = 6
BORDER_BR = 7

l = doglcd.DogLCD(21,22,24,18,-1,-1)
r = doglcd.DogLCD(21,22,24,23,-1,-1)

_ = 32


buf_l = [
	4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
	2,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,
	6,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]

buf_r = [
	0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5,
	_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,3,
	1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,7]


def initialise(disp):
	disp.begin(doglcd.DOG_LCD_M163, 0x28)
	disp.clear()
	disp.home()
	disp.noCursor()
	disp.noDoubleHeight()

	disp.createChar(BORDER_TOP,[31,31,0,0,0,0,0,0]) # Top

	disp.createChar(BORDER_TL,[31,31,24,24,24,24,24,24]) # Top Left
	disp.createChar(BORDER_BL,[24,24,24,24,24,24,31,31]) # Bottom Left

	disp.createChar(BORDER_TR,[31,31,3,3,3,3,3,3]) # Top Right
	disp.createChar(BORDER_BR,[3,3,3,3,3,3,31,31]) # Bottom Right
	
	disp.createChar(BORDER_LEFT,[24,24,24,24,24,24,24,24]) # Left
	disp.createChar(BORDER_RIGHT,[3,3,3,3,3,3,3,3]) # Right

	disp.createChar(BORDER_BOTTOM,[0,0,0,0,0,0,31,31]) # Bottom

	disp.home()


initialise(l)
initialise(r)


#for x in buf_l:
#	l.writeChar(x)
#for x in buf_r:
#	r.writeChar(x)

while 1:
	l.setCursor(0,0)
	r.setCursor(0,0)

	x = 1
	t = datetime.datetime.now().strftime("%H:%M:%S.%f")
	for letter in "This is a long piece of text which omg " + t + " such LCDs and wrap perfectly like magic oh wow and now overflow":
		if( x <= 96 ):
			if( math.ceil(float(x)/16.0) % 2 == 0 ):
				r.write(letter)
			else:
				l.write(letter)
		x+=1

while 1:
	l.setCursor(1,1)
	r.setCursor(0,1)
	x = 0
	for letter in "Hello World! It's " + datetime.datetime.now().strftime("%H:%M:%S.%f"):
		if( x < 30 ):
			if( x < 15):
				l.write(letter)
			else:
				r.write(letter)
			x+=1


examples/twin.py
#!/usr/bin/env python

import doglcd, datetime, time

# __init__(self, lcdSI, lcdCLK, lcdRS, lcdCSB, pin_reset, pin_backlight):
# pin   wpi   bcm
# SI      2   27 rev 2, 21 rev 1
# CLK     3   22
# RS      5   24
# CSB     4   23
# RST    -1
# BL     -1
#
lcd = doglcd.DogLCD(21,22,24,23,-1,-1)
lcdb = doglcd.DogLCD(21,22,24,18,-1,-1)

lcd.begin(doglcd.DOG_LCD_M163, 0x28)
lcd.clear()
lcd.home()
lcd.doubleHeight()
lcd.doubleHeightTop()

lcdb.begin(doglcd.DOG_LCD_M163, 0x28)
lcdb.clear()
lcdb.home()
lcdb.doubleHeight()
lcdb.doubleHeightBottom()

lcdb.noCursor()
lcd.noCursor()

lcdb.setCursor(1,0)
lcdb.write('DOGE!!!')

while 1:
	lcd.setCursor(0,0)
	lcdb.setCursor(0,0)
	t = datetime.datetime.now().strftime("%H:%M:%S.%f")
	lcd.write(t)
	lcdb.write(t)
	lcd.setCursor(0,1)
	d = datetime.datetime.now().strftime("%Y-%m-%d")
	lcd.write(d)
	

examples/sysinfo.py
import doglcd, psutil, time, datetime, math

import socket, fcntl, struct

def get_ip_address(ifname):
	try:
		s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		return socket.inet_ntoa(fcntl.ioctl(s.fileno(),
										0x8915, # SIOCGIFADDR
										struct.pack('256s', ifname[:15])
									)[20:24])
	except IOError:
		return ifname + ' not found!'

lcd = doglcd.DogLCD(21,22,24,18,-1,-1)

# Testing rig only, remove for production
# second LCD goes potty if you don't
# pull down its CSB pin!
alt = doglcd.DogLCD(21,22,24,23,-1,-1)
alt.begin(doglcd.DOG_LCD_M163, 0x28)
alt.clear()
alt.home()
alt.noCursor()
alt.noDoubleHeight()
alt.noAutoscroll()

lcd.begin(doglcd.DOG_LCD_M163, 0x28)
lcd.clear()
lcd.home()
lcd.noCursor()
lcd.noDoubleHeight()
lcd.noAutoscroll()

DOG_LCD_CHEVRON_L = 0b11111011
DOG_LCD_CHEVRON_R = 0b11111100

# We need some custom characters to
# complete the levels of bar graph
lcd.createChar(6,[0,31,0,0,0,0,0,0])
lcd.createChar(5,[0,0,31,0,0,0,0,0])
lcd.createChar(4,[0,0,0,0,31,0,0,0])
lcd.createChar(3,[0,0,0,0,0,31,0,0])
lcd.createChar(2,[0,0,0,0,0,0,0,31])

# Always call home after creating
# custom chars
lcd.home()

# Map out the levels of bar graph
graph_steps = [2,95,3,4,45,5,6,255]

graph_text = []

while 1:
	alt.setCursor(0,0)
	alt.write(get_ip_address('eth0'))

	alt.setCursor(0,1)
	alt.write(get_ip_address('wlan0'))

	ip = psutil.net_connections(kind='inet4')
	mem = psutil.virtual_memory()
	cpu = psutil.cpu_percent()

	lcd.setCursor(0,0)
	lcd.write("FREE " + str(mem.available/1024/1024) + 'MB/' + str(mem.total/1024/1024) + 'MB')

	lcd.setCursor(0,1)
	lcd.write("CPU% " + str(round(cpu,2)) + '    ')

	idx = int(math.floor((7.0/100.0) * cpu))
	
	graph_text.append(graph_steps[idx])

	if len(graph_text) > 16:
		graph_text.pop(0)

	lcd.setCursor(0,2)
	for char in graph_text:
		lcd.writeChar(char)

	time.sleep(0.25)

doglcd.py
#!/usr/bin/env python

import atexit, time
import RPi.GPIO as GPIO

DOG_LCD_M081 = 1
DOG_LCD_M162 = 2
DOG_LCD_M163 = 3

DOG_LCD_COMMAND_CLEAR = 0x01
DOG_LCD_COMMAND_HOME = 0x02

DOG_LCD_CURSOR_ON  = 0x02
DOG_LCD_CURSOR_OFF = 0x00

DOG_LCD_DISPLAY_ON = 0x04
DOG_LCD_DISPLAY_OFF = 0x00

DOG_LCD_BLINK_OFF = 0x00
DOG_LCD_BLINK_ON = 0x01

DOG_LCD_SCROLL_LEFT = 0x18
DOG_LCD_SCROLL_RIGHT = 0x1C



class DogLCD:
	def __init__(self, lcdSI, lcdCLK, lcdRS, lcdCSB, pin_reset, pin_backlight):
		GPIO.setmode(GPIO.BCM)
		GPIO.setwarnings(False)

		self.lcdSI = lcdSI
		self.lcdCLK = lcdCLK
		self.lcdRS = lcdRS
		self.lcdCSB = lcdCSB
		self.lcdReset = pin_reset
		self.lcdBacklight = pin_backlight

		self.animations = [[],[],[],[],[],[],[],[]]

	def begin(self, model, contrast):

		self.model = model
		self.startAddress = [-1,-1,-1]

		if (model==DOG_LCD_M081):
			self.rows = 1
			self.cols = 8
			self.memSize = 80
			self.startAddress[0] = 0
			self.startAddress[1] = -1
			self.startAddress[2] = -1
		elif (model==DOG_LCD_M162):
			self.rows = 2
			self.cols = 16
			self.memSize = 40
			self.startAddress[0] = 0
			self.startAddress[1] = 0x40
			self.startAddress[2] = -1
		elif (model==DOG_LCD_M163):
			self.rows = 3
			self.cols = 16
			self.memSize = 16
			self.startAddress[0] = 0
			self.startAddress[1] = 0x10
			self.startAddress[2] = 0x20
		else:
			# Unknown or unsupported model
			return False

		if(self.lcdReset != -1):
			GPIO.setup(self.lcdReset, GPIO.OUT)
			GPIO.output(self.lcdReset, GPIO.HIGH)

		if(self.lcdBacklight != -1):
			GPIO.setup(self.lcdBacklight, GPIO.OUT)
			GPIO.output(self.lcdBacklight, GPIO.LOW)

		if(contrast < 0 or contrast > 0x3F):
			# Contrast falls outside valid range
			return False

		self.contrast = contrast

		GPIO.setup(self.lcdCSB,GPIO.OUT);
		GPIO.output(self.lcdCSB,GPIO.HIGH);
		GPIO.setup(self.lcdSI,GPIO.OUT);
		GPIO.output(self.lcdSI,GPIO.HIGH);
		GPIO.setup(self.lcdCLK,GPIO.OUT);
		GPIO.output(self.lcdCLK,GPIO.HIGH);
		GPIO.setup(self.lcdRS,GPIO.OUT);
		GPIO.output(self.lcdRS,GPIO.HIGH);

		self.reset()
		return True

	def reset(self):
		if( self.lcdReset != -1):
			GPIO.output(self.lcdReset,GPIO.LOW)
			self.delay(40)
			GPIO.output(self.lcdReset,GPIO.HIGH)
			self.delay(40)
		else:
			self.delay(50)

		if(self.model==DOG_LCD_M081):
			self.instructionSetTemplate=0x30
		elif(self.model==DOG_LCD_M162):
			self.instructionSetTemplate=0x38
		elif(self.model==DOG_LCD_M163):
			self.instructionSetTemplate=0x38

		self.setInstructionSet(1)

		self.writeCommand(0x1D,30)

		self.setContrast(self.contrast)

		self.displayMode 	= DOG_LCD_DISPLAY_ON
		self.cursorMode 		= DOG_LCD_CURSOR_ON
		self.blinkMode 		= DOG_LCD_BLINK_OFF
		self.writeDisplayMode()
		self.entryMode 		= 0x04
		self.clear()
		self.leftToRight()

	def setContrast(self, contrast):
		if(contrast<0 or contrast>0x3F):
			return False

		# For 3.3v operation the booster must be on, which is
		# on the same command as the (2-bit) high-nibble of contrast
		self.writeCommand( (0x54 | ((self.contrast >> 4) & 0x03)), 30 )

		self.writeCommand( 0x6B, 30 )
		# Set low-nibble of the contrast
		self.writeCommand( (0x70 | (contrast & 0x0F) ), 30 )

	def clear(self):
		self.writeCommand(DOG_LCD_COMMAND_CLEAR,1080)

	def home(self):
		self.writeCommand(DOG_LCD_COMMAND_HOME,1080)

	def setCursor(self, col, row):
		if( col > self.memSize or row >= self.rows ):
			# Cursor outside of valid range
			return False
		address = (self.startAddress[row]+col) & 0x7F
		self.writeCommand(0x80|address,30)

	def noDisplay(self):
		self.displayMode=DOG_LCD_DISPLAY_OFF #0x00
		self.writeDisplayMode()

	def display(self):
		self.displayMode=DOG_LCD_DISPLAY_ON #0x04
		self.writeDisplayMode()

	def noCursor(self):
		self.cursorMode=DOG_LCD_CURSOR_OFF #0x00
		self.writeDisplayMode()

	def cursor(self):
		self.cursorMode=DOG_LCD_CURSOR_ON #0x02
		self.writeDisplayMode()

	def noBlink(self):
		self.blinkMode=DOG_LCD_BLINK_OFF
		self.writeDisplayMode()

	def blink(self):
		self.blinkMode=DOG_LCD_BLINK_ON
		self.writeDisplayMode()

	def scrollDisplayLeft(self):
		self.setInstructionSet(0)
		self.writeCommand(DOG_LCD_SCROLL_LEFT,30) # 0x18

	def scrollDisplayRight(self):
		self.setInstructionSet(0)
		self.writeCommand(DOG_LCD_SCROLL_RIGHT,30) # 0x1C

	def leftToRight(self):
		self.entryMode|=0x02
		self.writeCommand(self.entryMode,30)

	def rightToLeft(self):
		self.entryMode&=~0x02
		self.writeCommand(self.entryMode,30)

	def autoScroll(self):
		self.entryMode|=0x01
		self.writeCommand(self.entryMode,30)

	def noAutoscroll(self):
		self.entryMode&=~0x01
		self.writeCommand(self.entryMode,30)

	def doubleHeight(self):
		self.writeCommand(0b00100110,30)

	def noDoubleHeight(self):
		self.writeCommand(0b00100001,30)

	def doubleHeightTop(self):
		self.writeCommand(0b00011000,30)

	def doubleHeightBottom(self):
		self.writeCommand(0b00010000,30)

	def createAnimation(self, anim_pos, anim_map, frame_rate):
		self.createChar(anim_pos, anim_map[0])
		self.animations[anim_pos] = [anim_map,frame_rate]
		self.setCursor(0,1)
	
	def updateAnimations(self):
		for i,animation in enumerate(self.animations):
			if len(animation) == 2:
				anim = animation[0]
				fps = animation[1]
				frame = anim[ int(round(time.time()*fps) % len(anim)) ]
				self.createChar(i,frame)
		self.setCursor(0,1)

	def createChar(self, char_pos, char_map):
		if(char_pos<0 or char_pos>7):
			return False

		baseAddress = char_pos*8
		self.setInstructionSet(0)
		for i in range(0,8):
			self.writeCommand((0x40|(baseAddress+i)),30)
			self.writeChar(char_map[i])
		self.setInstructionSet(0)
		self.writeDisplayMode()

	def writeDisplayMode(self):
		self.writeCommand((0x08 | self.displayMode | self.cursorMode | self.blinkMode), 30)

	def setBacklight(self, value, usePWM):
		if(self.lcdBacklight!=-1 and value>=0):
			if(value == LOW):
				GPIO.output(self.lcdBacklight,GPIO.LOW)
			else:
				GPIO.output(self.lcdBacklight,GPIO.HIGH)

	def setInstructionSet(self, instruction_set):
		if( instruction_set < 0 or instruction_set > 3 ):
			return False
		self.writeCommand( self.instructionSetTemplate | instruction_set, 30 )

	def writeChar(self, value):
		# Switch to data mode
		GPIO.output( self.lcdRS, GPIO.HIGH )
		self.spiTransfer(value,30)
		return True

	def writeCommand(self, value, delay):
		# Switch to command-mode
		GPIO.output( self.lcdRS, GPIO.LOW )
		self.spiTransfer(value,delay)

	def spiTransfer(self, value, delay):
		# Ensure Clock starts high
		GPIO.output(self.lcdCLK,GPIO.HIGH)

		GPIO.output(self.lcdCSB,GPIO.LOW)

		for i in reversed(range(0,8)):
			GPIO.output( self.lcdSI, value & (1 << i) )
			GPIO.output( self.lcdCLK, GPIO.LOW )
			GPIO.output( self.lcdCLK, GPIO.HIGH )

		GPIO.output(self.lcdCSB,GPIO.HIGH)

		# Wait for command to complete
		self.delayMicroseconds( delay )

	def write(self, string):
		string = str( string )
		for char in string:
			self.writeChar(ord(char))

	def delayMicroseconds(self, delay):
		time.sleep(delay/1000000)

	def delay(self, delay):
		time.sleep(delay/1000)

atexit.register(lambda: GPIO.cleanup())